File: patternlets/MPI/12.reduction/reduction.c
/* reduction.c
* ... illustrates the use of MPI_Reduce()...
* Joel Adams, Calvin College, November 2009.
*
* Usage: mpirun -np N ./reduction
*
* Exercise:
* - Compile and run, varying N: 4, 6, 8, 10.
* - Explain behavior of MPI_Reduce().
*/

#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char** argv) {
    int numProcs = -1, myRank = -1, square = -1, max = -1, sum = 0;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &numProcs);
    MPI_Comm_rank(MPI_COMM_WORLD, &myRank);

    square = (myRank+1) * (myRank+1);

    printf("Process %d computed %d\n", myRank, square);

    MPI_Reduce(&square, &sum, 1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);

    MPI_Reduce(&square, &max, 1, MPI_INT, MPI_MAX, 0, MPI_COMM_WORLD);

    if (myRank == 0) {
        printf("\nThe sum of the squares is %d\n\n", sum);
        printf("The max of the squares is %d\n\n", max);
    }

    MPI_Finalize();

    return 0;
}

File: patternlets/java-OpenMPI/13.reduction/Reduction.java
/* Reduction.java
 * ... illustrates how to use the reduction pattern 
 *     (which combines distributed values in O(lg(P)) time)
 *      using OpenMPI's Java interface.
 *
 * Joel Adams, Calvin University, November 2019.
 *
 * Usage: mpirun -np 4 java ./Reduction
 *
 * Exercise:
 * - Compile and run, varying N = 1, 2, 3, 4, 6, 8, 10.
 * - Explain behavior of the reduce operation.
 */

import mpi.*;
import java.nio.IntBuffer;

public class Reduction {

 public static void main(String [] args) throws MPIException {
    MPI.Init(args);

    Comm comm        = MPI.COMM_WORLD;
    int id           = comm.getRank();
    int numProcesses = comm.getSize();

    int square       = (id+1) * (id+1);
    IntBuffer squareBuffer = MPI.newIntBuffer(BUFFER_SIZE);
    squareBuffer.put(square);

    IntBuffer sumSquaresBuffer = MPI.newIntBuffer(BUFFER_SIZE);
    comm.reduce(squareBuffer, sumSquaresBuffer, BUFFER_SIZE,
                 MPI.INT, MPI.SUM, MASTER);

    IntBuffer maxBuffer = MPI.newIntBuffer(BUFFER_SIZE);
    comm.reduce(squareBuffer, maxBuffer, BUFFER_SIZE,
                 MPI.INT, MPI.MAX, MASTER);

    if ( id == MASTER) {
        String squareMsg = "\nThe sum of the squares from 1 to "
                            + numProcesses + " is " 
                            + sumSquaresBuffer.get(0) + "\n\n";
        String maxMsg    = "The max of the squares from 1 to "
                            + numProcesses + " is " 
                            + maxBuffer.get(0) + "\n\n";
        System.out.print(squareMsg);
        System.out.print(maxMsg);
    }

    MPI.Finalize();
  }

  private static int BUFFER_SIZE = 1;
  private static int MASTER      = 0;
}

